asab/metrics/prometheus.py

```diff
@@ -0,0 +1,112 @@
+import asab
+import re
+import logging
+
+
+L = logging.getLogger(__name__)
+
+
+def validate_format(name):
+    name = name.lower()
+    regex = r"[\WÁ-ž]+"
+    subst = "_"
+    result = re.sub(regex, subst, name)
+    result = result.lstrip("_")
+    if result.endswith("total") or result.endswith("created"):
+        L.warning("Invalid OpenMetrics format. Name MUST NOT end with total or created.")
+        result = result.rstrip("total")
+        result = result.rstrip("created")
+    return result
+
+
+def validate_value(value):
+    return isinstance(value, [int, float])
+
+
+def get_labels(tags):
+    labels_str = "{"
+    for tag in tags.keys():
+        if tag == "host" or tag == "unit" or tag == "help":
+            continue
+        else:
+            if tag.startswith("_"):
+                tag = tag.lstrip("_")
+            labels_str += '{}="{}",'.format(tag, tags.get(tag))
+    labels_str = labels_str.rstrip(",")
+    labels_str += "}"
+    if labels_str == "{}":
+        return None
+    else:
+        return labels_str
+
+
+def metric_to_text(metric, type):
+    metric_lines = []
+    m_name = metric.get("Name")
+    tags = metric.get("Tags")
+    labels_str = get_labels(tags)
+
+    for v_name, value in metric.get("Values").items():
+        if validate_value is False:
+            L.warning("Invalid OpenMetrics format. Value must be float or integer.")
+            continue
+        else:
+            name = "_".join([m_name, v_name])
+            name = validate_format(name)
+            # If a unit is specified it MUST be provided in a UNIT metadata line. In addition, an underscore and the unit MUST be the suffix of the MetricFamily name.
+            if tags.get("unit"):
+                unit = tags.get("unit")
+                unit = validate_format(unit)
+                name += "_{}".format(unit)
+                metric_lines.append("# TYPE {} {}".format(name, type))
+                metric_lines.append("# UNIT {} {}".format(name, unit))
+            else:
+                unit = None
```

Comment:
`unit = "total"` ... or based on the type of Metrics.

 if I understand OpenMetrics/Prometheus doc correctly

Reply:
As I understand it, "_total" is not a unit - it is an obligatory sufix for counter metrics.

Reply:
My understanding is that it is used when then "unit" is not present

Reply:
Ok, that's correct